eliminate garbage when `SkipValue` encounters strings

bobby-stripe opened the pull request:
Thanks for the great library!  When using this to parse a large (10+MB) amount of JSON, we see a significant (the majority! after https://github.com/launchdarkly/go-jsonstream/pull/19 ) amount of garbage created when `ObjectState.Next()` calls `Reader.SkipValue()`:
<img width="350" alt="image" src="https://github.com/launchdarkly/go-jsonstream/assets/35409959/77f036c7-b87d-4867-ac0b-0167c8b538cd">

With this PR an internal benchmark shows an 80% reduction in allocation count (with a more modest reduction in MB allocated, although a majority of the remaining allocs are just the JSON file contents we are operating on):

```
name             old time/op    new time/op    delta
LoadFromFile-10     108ms ± 1%     106ms ± 1%   -1.16%  (p=0.008 n=5+5)

name             old alloc/op   new alloc/op   delta
LoadFromFile-10    26.0MB ± 0%    24.9MB ± 0%   -3.90%  (p=0.008 n=5+5)

name             old allocs/op  new allocs/op  delta
LoadFromFile-10     34.0k ± 0%      5.8k ± 0%  -82.80%  (p=0.029 n=4+4)
```

I've added a test as well that fails without this change applied, and with this change applied shows that streaming over an object with string values is 0 allocation.

Happy to iterate if you have feedback!

bobby-stripe committed:
jreader: when SkipValue is called, don't allocate strings that immediately become garbage

bobby-stripe committed:
jreader: add test ensuring that SkipValue doesn't allocate strings

cwaldren-ld commented:
Hi @bobby-stripe , thank you for doing this work - we definitely appreciate changes like this. I can't carefully review at this moment, but I'll let you know if I have any questions when I circle back.

_Filed internally as 224240_.

bobby-stripe commented:
@cwaldren-ld sounds great, thanks!

bobby-stripe commented:
I pushed an updated set of commits that builds + tests correctly with both the default + easyjson builds

bobby-stripe commented:
@cwaldren-ld gentle ping, any update on this?  There is an open question on the other PR, but I think this PR is ready to review

cwaldren-ld commented on jreader/token_reader_default.go:
```diff
@@ -17,6 +17,8 @@ import (
 	"unicode/utf8"
 )
 
```
This constant and the one in `jreader/token_reader_easyjson.go` could use a comment explaining that they're used in the tests

bobby-stripe replied:
@cwaldren-ld I pushed an updated commit that has this comment!

cwaldren-ld reviewed (commented):
Hey @bobby-stripe , sorry for the delay. This is looking good, just had one comment.

Edit:

Running CI here https://github.com/launchdarkly/go-jsonstream/pull/21
```
# github.com/launchdarkly/go-jsonstream/v3/jreader
jreader/reader.go:194:19: r.tr.StringAsBytes undefined (type tokenReader has no field or method StringAsBytes)
```
(can repro locally with `make test-easyjson`)

bobby-stripe commented:
> Hey @bobby-stripe , sorry for the delay. This is looking good, just had one comment.

No problem!  Addressed the comment with a force push to this branch - you will likely want to re-do the test merge of it into `contrib`.

> 
> Edit:
> 
> Running CI here #21
> 
> ```
> # github.com/launchdarkly/go-jsonstream/v3/jreader
> jreader/reader.go:194:19: r.tr.StringAsBytes undefined (type tokenReader has no field or method StringAsBytes)
> ```
> 
> (can repro locally with `make test-easyjson`)

sorry for this -- this was a problem with the other PR.  I put up a fix for that here: https://github.com/launchdarkly/go-jsonstream/pull/22

bobby-stripe commented:
going to close this as its merged into `contrib` already
